stan/math/rev/core/var.hpp

```diff
@@ -415,6 +433,68 @@ class var_value {
     }
     return os << v.val();
   }
+
+  /**
+   * Assignment of another plain var value, when this also contains a plain
+   * type.
+   * @tparam S type of the value in the `var_value` to assing
+   * @param other the value to assign
+   * @return this
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
+            require_all_plain_type_t<T, S>* = nullptr>
+  var_value<T> operator=(const var_value<S>& other) {
+    vi_ = other.vi_;
+    return *this;
+  }
+
+  /**
+   * Assignment of another var value, when either this or the other one does not
+   * contain a plain type.
+   * @tparam S type of the value in the `var_value` to assing
+   * @param other the value to assign
+   * @return this
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr,
+            require_any_not_plain_type_t<T, S>* = nullptr>
+  var_value<T> operator=(const var_value<S>& other) {
+    arena_t<plain_type_t<T>> prev_val = vi_->val_;
+    vi_->val_ = other.val();
+    // no need to change any adjoints - these are just zeros before the reverse
+    // pass
+
+    reverse_pass_callback(
+        [this_vi = this->vi_, other_vi = other.vi_, prev_val]() mutable {
+          this_vi->val_ = prev_val;
+
+          // we have no way of detecting aliasing between this->vi_->adj_ and
+          // other.vi_->adj_, so we must copy adjoint before reseting to zero
+
+          // we can reuse prev_val instead of allocating a new matrix
+          prev_val = this_vi->adj_;
+          this_vi->adj_.setZero();
+          other_vi->adj_ += prev_val;
+        });
+    return *this;
+  }
+
+  // this is a bit ugly workarount to allow var_value<double> to have default
+  // assignment operator, which avoids warnings when used in Eigen matrices.
+ private:
+  struct not_var_value {};
+
+ public:
+  /**
+   * Copy assignment operator delegates to general assignment operator if the
+   * var_value contains eigen type.
+   * @param other the value to assing
+   * @return this
+   */
+  var_value<T> operator=(std::conditional_t<is_eigen<value_type>::value,
+                                            const var_value<T>&, not_var_value>
+                             other) {
+    return operator=<T>(other);
```

Comment:
Looking over all this I think we should go back to having another template parameter for `var_value` to do SFINAE on. eod we have a lot of stuff here that is just for eigen types and some stuff that's just for doubles so I think it would make more sense. There would be some code duplication but I think things will look a lot cleaner

```
template <typename T, typename = void>
class var_value;

template <typename T>
class var_value<T, require_floating_point_t<T>> {
  // double stuff  
}

template <typename T>
class var_value<T, require_eigen_t<T>> {
  // eigen stuff  
}
```

Reply:
I agree. Separate PR though.